src/core/hle/kernel/thread.cpp

```diff
@@ -3,433 +3,972 @@
-    thread_managers[processor_id]->ready_queue.prepare(priority);
+void Thread::Stop() {
+    if (status != Thread::Status::Destroyed) {
+        status = Thread::Status::Destroyed;
 
-    thread->thread_id = NewThreadId();
-    thread->status = ThreadStatus::Dormant;
-    thread->entry_point = entry_point;
-    thread->stack_top = stack_top;
-    thread->nominal_priority = thread->current_priority = priority;
-    thread->last_running_ticks = timing.GetGlobalTicks();
-    thread->processor_id = processor_id;
-    thread->wait_objects.clear();
-    thread->wait_address = 0;
-    thread->name = std::move(name);
-    thread_managers[processor_id]->wakeup_callback_table[thread->thread_id] = thread.get();
-    thread->owner_process = owner_process;
+        for (auto mutex : held_mutexes) {
+            mutex->Release(this);
+        }
+        held_mutexes.clear();
 
-    // Find the next available TLS index, and mark it as used
-    auto& tls_slots = owner_process->tls_slots;
+        core.threads.erase(std::remove(core.threads.begin(), core.threads.end(), this),
+                           core.threads.end());
 
-    auto [available_page, available_slot, needs_allocation] = GetFreeThreadLocalSlot(tls_slots);
+        // Mark the TLS slot in the thread's page as free.
+        u32 tls_page = (tls_address - Memory::TLS_AREA_VADDR) / Memory::PAGE_SIZE;
+        u32 tls_slot =
+            ((tls_address - Memory::TLS_AREA_VADDR) % Memory::PAGE_SIZE) / Memory::TLS_ENTRY_SIZE;
+        process->tls_slots[tls_page].reset(tls_slot);
+    }
+}
 
-    if (needs_allocation) {
-        // There are no already-allocated pages with free slots, lets allocate a new one.
-        // TLS pages are allocated from the BASE region in the linear heap.
-        auto memory_region = GetMemoryRegion(MemoryRegion::BASE);
+bool Thread::IsWokenBy(const Kernel::WaitObject* object) {
 
-        // Allocate some memory from the end of the linear heap for this region.
-        auto offset = memory_region->LinearAllocate(Memory::PAGE_SIZE);
-        if (!offset) {
-            LOG_ERROR(Kernel_SVC,
-                      "Not enough space in region to allocate a new TLS page for thread");
-            return ERR_OUT_OF_MEMORY;
+    switch (status) {
+    case Status::WaitSyncAny:
+    case Status::WaitHleEvent:
+    case Status::WaitIPC:
+        return std::find_if(waiting_on.begin(), waiting_on.end(), [&](const auto& obj) {
+                   return obj.get() == object;
+               }) != waiting_on.end();
+
+    case Status::WaitSyncAll:
+        return std::find_if(waiting_on.begin(), waiting_on.end(), [&](const auto& obj) {
+                   return obj.get() == object || !obj->ShouldWait(this);
+               }) == waiting_on.end();
+
+    default:
+        return false;
+    }
+}
+
+void Thread::ResumeFromWait(Kernel::WaitObject* object) {
+    switch (status) {
+
+    case Status::WaitSyncAll:
+        for (auto obj : waiting_on) {
```

Comment:
```suggestion
        for (auto& obj : waiting_on) {
```

otherwise this can cause atomic reference count increment/decrement churn.